### PR TITLE
Adding support for feature omnistack_clusters throughput

### LINKED ISSUE #1: all unit test cases are not being executed via tox
There appears to be a total of 106 unique unit test cases within the ./tests/unit directory structure

`python -m unittest discover` executes the 34 test cases directory under ./tests/unit, but skips the 63 tests that are under ./tests/unit/resources.  

out of the 63 tests that are under ./tests/unit/resources, 17 currently fail

The solution is to create an empty file called init.py to ./tests/unit/resources directory in order to allow the unittest discover routine to find the test cases

### PR DESCRIPTION
**Pull request by KalpanaBhardwaj**
Signed-off-by: KalpanaBhardwaj <bhardwaj@hpe.com>

### Description
Adding support for feature omnistack_clusters throughput

### Issues Resolved
Defect fix for adding filters in do_get method  for PR #66 

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass(`$ tox`).
- [x] New functionality has been documented in the README if applicable.
  - [x] New functionality has been thoroughly documented in the examples (please include helpful comments).
  - [x] New endpoints supported are updated in the endpoints-support.md file.
- [x] Changes are documented in the CHANGELOG.

### CODE REVIEW

File: simplivity/resources/resource.py [hunk 1]
```diff
@@ -212,15 +212,19 @@ def task_affected_resources(self, task, timeout):
 
         return affected_resources
 
-    def do_get(self, uri):
+    def do_get(self, uri, filters=None):
         """Makes get requests
 
         Args:
             uri: URI of the resource
+            flags: Dictionary of filters, example: {'name': 'name'}
```
Comment (svt-hchaudhari) [inline simplivity/resources/resource.py]: Variable name is different in the doc section then the name passed in the method

**Review by svt-hchaudhari — CHANGES_REQUESTED** [review]
Variable name is not matching with name passed in the method should be filters

**Comment by svt-hchaudhari** [comment]
@KalpanaBhardwaj, Need update in the CHANGELOG.md for the issue addressed in PR

**Comment by KalpanaBhardwaj** [comment]
> Variable name is not matching with name passed in the method should be filters
Thanks Hemant, the suggestion commit was already in progress. Its done.

File: endpoints-support.md [hunk 2]
```diff
@@ -32,10 +32,13 @@ Refer SimpliVity REST API doc for the resource endpoints documentation [HPE Simp
 |<sub>/hosts/{hostId}/shutdown_virtual_controller </sub>           |POST    |
 |<sub>/hosts/{hostId}/virtual_controller_shutdown_status</sub>     |GET     |
 |     **OmniStack Clusters**
-|<sub>/omnistack_clusters</sub>                                    |GET     |
-|<sub>/omnistack_clusters/time_zone_list</sub>                     |GET     |
-|<sub>/omnistack_clusters/{clusterId}/connected_clusters</sub>     |GET     |
-|<sub>/omnistack_clusters/{clusterId}/set_time_zone</sub>          |POST    |
+
```
Comment (sijeesh) [inline endpoints-support.md]: Incorrect formatting, delete this line

File: endpoints-support.md [hunk 3]
```diff
@@ -32,10 +32,13 @@ Refer SimpliVity REST API doc for the resource endpoints documentation [HPE Simp
...
+|<sub>/omnistack_clusters	</sub>                                                        |GET       |
+|<sub>/omnistack_clusters/time_zone_list  </sub>                                          |GET       |
+|<sub>/omnistack_clusters/{clusterId}/connected_clusters  </sub>                          |GET       |
+|<sub>/omnistack_clusters/{clusterId}/set_time_zone </sub>                                |POST      |
+|<sub>/omnistack_clusters/{clusterId}/throughput </sub>                                   |GET       |
+
```
Comment (sijeesh) [inline endpoints-support.md]: delete this line as well.

File: simplivity/resources/omnistack_clusters.py [hunk 4]
```diff
@@ -130,9 +130,27 @@ def get_connected_clusters(self):
...
+            range : A range in seconds (the duration from the specified point in time) Default: 43200
+
+        Returns:
+            cluster_throughput object.
+        """
+        method_url = "{}/{}/throughput".format(URL, self.data["id"])
+        filters = {}
+        if (destination_id):
+            filters['destination_id'] = destination_id
+        filters.update({'time_offset': time_offset, 'range': range})
```
Comment (sijeesh) [inline simplivity/resources/omnistack_clusters.py]: You can avoid this line by declaring filters dict with time_offset and range, then add destination_id if it's not null.

File: simplivity/resources/resource.py [hunk 5]
```diff
@@ -212,15 +212,19 @@ def task_affected_resources(self, task, timeout):
 
         return affected_resources
 
-    def do_get(self, uri):
+    def do_get(self, uri, filters=None):
```
Comment (sijeesh) [inline simplivity/resources/resource.py:215]: I think these kinds of changes should go as separate PRs

**Review by sijeesh — CHANGES_REQUESTED** [review]

**Comment by KalpanaBhardwaj** [comment]
> @KalpanaBhardwaj, Need update in the CHANGELOG.md for the issue addressed in PR

Its done.

File: simplivity/resources/omnistack_clusters.py [hunk 6]
```diff
@@ -130,9 +130,27 @@ def get_connected_clusters(self):
         connected_clusters = self._client.do_get(method_url).get("omnistack_clusters", [])
 
         clusters = [self._clusters.get_by_id(cluster["id"]) for cluster in connected_clusters]
-
         return clusters
 
+    def get_throughput(self, destination_id=None, time_offset=0, range=43200):
+        """Calculates the throughput between each pair of omnistack_clusters in the federation
+
+Args:
```
Comment (svt-hchaudhari) [inline simplivity/resources/omnistack_clusters.py]: The alignment of Args is incorrect, please correct the alignment.

File: simplivity/resources/omnistack_clusters.py [hunk 7]
```diff
@@ -130,9 +130,27 @@ def get_connected_clusters(self):
...
+            destination_id : The unique identifier (UID) of the omnistack_clusters to return
+            time_offset : A time offset in seconds (from now) or a datetime,
+                        expressed in ISO-8601 form, based on Coordinated Universal Time (UTC) Default: 0
+            range : A range in seconds (the duration from the specified point in time) Default: 43200
+
+        Returns:
+            cluster_throughput object.
```
Comment (svt-hchaudhari) [inline simplivity/resources/omnistack_clusters.py]: Please specify the type of the return object

File: tests/unit/resources/test_omnistack_clusters.py [hunk 8]
```diff
@@ -112,6 +112,20 @@ def test_get_connected_clusters(self, mock_get):
         mock_get.assert_has_calls([call('/omnistack_clusters/12345/connected_clusters'),
                                   call('/omnistack_clusters?case=sensitive&id=12345&limit=500&offset=0&order=descending&sort=name')])
 
+    @mock.patch.object(Connection, "get")
+    def test_get_throughput_without_filters(self, mock_get):
+        cluster_data = {'name': 'name1', 'id': '12345'}
+        cluster = self.clusters.get_by_data(cluster_data)
+        cluster.get_throughput()
```
Comment (svt-hchaudhari) [inline tests/unit/resources/test_omnistack_clusters.py]: Need to verify return object for the get throughput method

File: tests/unit/resources/test_omnistack_clusters.py [hunk 9]
```diff
@@ -112,6 +112,20 @@ def test_get_connected_clusters(self, mock_get):
...
+        mock_get.assert_has_calls([call('/omnistack_clusters/12345/throughput?range=43200&time_offset=0')])
+
+    @mock.patch.object(Connection, "get")
+    def test_get_throughput_with_filters(self, mock_get):
+        cluster_data = {'name': 'name1', 'id': '12345'}
+        cluster = self.clusters.get_by_data(cluster_data)
+        cluster.get_throughput("12345", 0, 200)
```
Comment (svt-hchaudhari) [inline tests/unit/resources/test_omnistack_clusters.py]: Same as above, we need to verify return type

File: CHANGELOG.md [hunk 10]
```diff
@@ -60,6 +61,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
     - resolve failing unit test cases
     - resolve the virtual machine `__refresh` method
     - url encode query string to support values that contain spaces
+    - resolve `do_get` method to append filters to uri
```
Comment (svt-hchaudhari) [inline CHANGELOG.md:71]: This change should be not part of this commit, it should be part of defect commit.

**Review by svt-hchaudhari — CHANGES_REQUESTED** [review]

File: endpoints-support.md (hunk 2 above)
Comment (KalpanaBhardwaj) [reply endpoints-support.md]: Done

File: endpoints-support.md (hunk 3 above)
Comment (KalpanaBhardwaj) [reply endpoints-support.md]: Done

File: simplivity/resources/omnistack_clusters.py (hunk 4 above)
Comment (KalpanaBhardwaj) [reply simplivity/resources/omnistack_clusters.py]: Updated

**Comment by KalpanaBhardwaj** [comment]
> @KalpanaBhardwaj, Need update in the CHANGELOG.md for the issue addressed in PR


Done.

File: simplivity/resources/omnistack_clusters.py (hunk 6 above)
Comment (KalpanaBhardwaj) [reply simplivity/resources/omnistack_clusters.py]: Updated the alignment correctly

File: simplivity/resources/omnistack_clusters.py (hunk 7 above)
Comment (KalpanaBhardwaj) [reply simplivity/resources/omnistack_clusters.py]: Updated as suggested

File: tests/unit/resources/test_omnistack_clusters.py (hunk 8 above)
Comment (KalpanaBhardwaj) [reply tests/unit/resources/test_omnistack_clusters.py]: Updated as suggested

File: tests/unit/resources/test_omnistack_clusters.py (hunk 9 above)
Comment (KalpanaBhardwaj) [reply tests/unit/resources/test_omnistack_clusters.py]: Updated as suggested

File: CHANGELOG.md (hunk 10 above)
Comment (KalpanaBhardwaj) [reply CHANGELOG.md:71]: Updated as suggested

File: simplivity/resources/resource.py (hunk 5 above)
Comment (KalpanaBhardwaj) [reply simplivity/resources/resource.py:215]: Hi @sijeesh ,
Since this enhancement was discovered during the implementation of current API, however the fix has dependency on overall working of this PR code. 
Even we thought of creating two different PRs however with this dependency the PR code can be merged and will not be completely functional hence we merged the code in same PR using two distinct commits. 

Please let us know your views on it, also if needed we can still go ahead and create two different PRs. 

Thank you looking forward to your views on it.

File: endpoints-support.md [hunk 11]
```diff
@@ -33,13 +33,14 @@ Refer SimpliVity REST API doc for the resource endpoints documentation [HPE Simp
 |<sub>/hosts/{hostId}/cancel_virtual_controller_shutdown</sub>     |POST    |
 |<sub>/hosts/{hostId}/remove_from_federation</sub>                 |POST    |
 |<sub>/hosts/{hostId}/hardware</sub>                               |GET     |
-|<sub>/hosts/{hostId}/shutdown_virtual_controller </sub>           |POST    |
+|<sub>/hosts/{hostId}/shutdown_virtual_controller</sub>           |POST    |
```
Comment (svt-kmcdaniel) [inline endpoints-support.md]: please line up the | with the others

Comment (KalpanaBhardwaj) [reply endpoints-support.md]: Thanks Ken, Corrected the alignment with others

**Review by sijeesh — APPROVED** [review]
LGTM
